constraints_file.py

```diff
@@ -22,14 +24,32 @@ def generate_shift_employee_combinations(employees: list[Employee], shifts: list
     return shift_combinations
 
 
-# A constraint that ensures that there will be exactly one employee in each shift per day
-def add_at_least_one_employee_per_shift_constraint(shifts: list[Shift], employees: list[Employee], constraint_model: cp_model.CpModel, shift_combinations: dict[FrozenShiftCombinationsKey, IntVar]) -> None:
+# A constraint that ensures that there will be exactly one employee in each shift per day.
+def add_exactly_one_employee_per_shift_constraint(shifts: list[Shift], employees: list[Employee], constraint_model: cp_model.CpModel, shift_combinations: dict[FrozenShiftCombinationsKey, IntVar]) -> None:
     for shift in shifts:
         all_employees_working_this_shift = []
 
         for employee in employees:
             key = FrozenShiftCombinationsKey(employee.employee_id, shift.shift_id)
-
             all_employees_working_this_shift.append(shift_combinations[key])
 
         constraint_model.AddExactlyOne(all_employees_working_this_shift)
+
+
+# A constraint that ensures that each employee works at most one shift per day.
+def add_at_most_one_shift_per_employee_in_the_same_day_constraint(shifts: list[Shift], employees: list[Employee], constraint_model: cp_model.CpModel, shift_combinations: dict[FrozenShiftCombinationsKey, IntVar]) -> None:
+    # Creating a dictionary that will hold a date as a key and all the shifts that starts in that date as values.
+    date_shifts_dict: dict[datetime.date, list[Shift]] = {}
+    list_of_dates = list(set(shift.start_time.date() for shift in shifts))
+
+    for date in list_of_dates:
+        date_shifts_dict[date] = [shift for shift in shifts if shift.start_time.date() == date]
+
+    for shifts_in_day in date_shifts_dict.values():
```

Comment:
you wrote alot of code considering all you want is to get all the shifts, grouped by their day
https://docs.python.org/3/library/itertools.html#itertools.groupby

its a one liner with a lambda